### ISSUE TITLE
Allow in_array check to constrain type of array

### ISSUE THREAD

**Issue by bdsl** [vimeo/psalm#2561]
If I have a varable of type `array<string|null>`, and I check that `\in_array(null, $array)` returns false, I would like the type of the array to be updated to `array<string>`, i.e. I would like the follow have no errors: https://psalm.dev/r/8999dd6ac4

Various workarounds are possible, such as moving the throw into a closure passed to array_map, and using the array returned from array map. It will be identical to the input array, but psalm constrains only the type of the return value: https://psalm.dev/r/b8c536ceb6

**Comment by muglug** [comment]
I'm not sure about this. It's somewhat esoteric, to me at least – I haven't seen that pattern used before, and it's also a little dangerous without `in_array` strict checks: https://3v4l.org/GsjkM

**Comment by muglug** [comment]
I'd encourage you to try adding this yourself.

It'll require two new types of assertion – I'd name them `array-has-falsy` and `array-has-null` to correspond to `in_array(null, $arr)` and `in_array(null, $arr, true)` respectively (you'd detect those in `AssertionFinder::processFunctionCall` I think).

Then you can add code to handle them in `AssertionReconciler` and `NegatedAssertionReconciler`, respectively.

Let me know if you have questions!

**Comment by bdsl** [comment]
Good point about the danger - I keep forgetting the third parameter to [in_array](https://www.php.net/manual/en/function.in-array.php).

In the particular use case I had for this I noticed that the array is an array of entities generated from an array_map over a list of entity IDs, so I realised I could just add the null check into the closure that was already there.

I'll think about if I have time to add the assertions - might be interesting since I haven't done anything in that area of Psalm.

**Comment by muglug** [comment]
[Here's a recent commit](https://github.com/vimeo/psalm/commit/f97a8f0d5bb99e0aa1c959e62a7998d791eeb57a#diff-307e7139f4685a7488fe37638fe73cf2) where I added a somewhat-similar check, though that implementation is a tad more complicated than this needs to be.

**Comment by bdsl** [comment]
Thanks, that looks helpful.
